src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,1045 @@
+    return Type::WidenMin;
+  }
+  if (_srange._lo == std::numeric_limits<S>::min() && _srange._hi == std::numeric_limits<S>::max() &&
+      _urange._lo == std::numeric_limits<U>::min() && _urange._hi == std::numeric_limits<U>::max() &&
+      _bits._zeros == U(0) && _bits._ones == U(0)) {
+    // bottom type
+    return Type::WidenMax;
+  }
+  return widen;
+}
+
+#ifdef ASSERT
+template <class S, class U>
+bool TypeIntPrototype<S, U>::contains(S v) const {
+  U u(v);
+  return v >= _srange._lo && v <= _srange._hi && u >= _urange._lo && u <= _urange._hi && _bits.is_satisfied_by(u);
+}
+
+// Verify that this set representation is canonical
+template <class S, class U>
+void TypeIntPrototype<S, U>::verify_constraints() const {
+  // Assert that the bounds cannot be further tightened
+  assert(contains(_srange._lo) && contains(_srange._hi) &&
+         contains(S(_urange._lo)) && contains(S(_urange._hi)), "");
+
+  // Assert that the bits cannot be further tightened
+  if (U(_srange._lo) == _urange._lo) {
+    assert(!adjust_bits_from_bounds(_bits, _urange)._progress, "");
+  } else {
+    RangeInt<U> neg_range{U(_srange._lo), _urange._hi};
+    auto neg_bits = adjust_bits_from_bounds(_bits, neg_range);
+    assert(neg_bits._present, "");
+    assert(!adjust_bounds_from_bits(neg_range, neg_bits._result)._progress, "");
+
+    RangeInt<U> pos_range{_urange._lo, U(_srange._hi)};
+    auto pos_bits = adjust_bits_from_bounds(_bits, pos_range);
+    assert(pos_bits._present, "");
+    assert(!adjust_bounds_from_bits(pos_range, pos_bits._result)._progress, "");
+
+    assert((neg_bits._result._zeros & pos_bits._result._zeros) == _bits._zeros &&
+           (neg_bits._result._ones & pos_bits._result._ones) == _bits._ones, "");
+  }
+}
+#endif // ASSERT
+
+template class TypeIntPrototype<jint, juint>;
+template class TypeIntPrototype<jlong, julong>;
+template class TypeIntPrototype<intn_t<1>, uintn_t<1>>;
+template class TypeIntPrototype<intn_t<2>, uintn_t<2>>;
+template class TypeIntPrototype<intn_t<3>, uintn_t<3>>;
+template class TypeIntPrototype<intn_t<4>, uintn_t<4>>;
+
+// Compute the meet of 2 types, when dual is true, we are actually computing the
+// join.
+template <class CT, class S, class U>
+const Type* TypeIntHelper::int_type_xmeet(const CT* i1, const Type* t2, make_type_t<S, U> make, bool dual) {
+  // Perform a fast test for common case; meeting the same types together.
+  if (i1 == t2 || t2 == Type::TOP) {
+    return i1;
+  }
+  const CT* i2 = t2->try_cast<CT>();
+  if (i2 != nullptr) {
+    if (!dual) {
+    // meet
+      return make(TypeIntPrototype<S, U>{{MIN2(i1->_lo, i2->_lo), MAX2(i1->_hi, i2->_hi)},
+                                         {MIN2(i1->_ulo, i2->_ulo), MAX2(i1->_uhi, i2->_uhi)},
+                                         {i1->_bits._zeros & i2->_bits._zeros, i1->_bits._ones & i2->_bits._ones}},
+                  MAX2(i1->_widen, i2->_widen), false);
```

Comment:
Ok, this looks like a union. And below like a intersection.

Reply:
Why do you handle the `widen` differently here?

Reply:
Because the join of 2 widens should be the smaller value and the meet of 2 widens should be the larger one.